### ISSUE TITLE
Node to Auth connection count increase

### ISSUE THREAD

**Issue by russjones** [gravitational/teleport#17023]
Teleport 9.3.x nodes maintained a single connection to Auth service. 

tarting in Teleport 10.x.x nodes started maintaining 3 connections to Auth service.

Investigate if the increase in number of connections is intentional and why.

**Comment by zmb3** [comment]
This is due to the inventory control stream introduced in v10. By v12 we should be able to get back down to a single auth client.

**Comment by fspmarshall** [comment]
The fix here is basically to use the `Instance` client for all auth client operations, rather than allocating a separate client for each service.

The `Instance` client's cert claims all services, so it aught to be a drop-in replacement, but there may be edge cases where we'll need to fix some checks.

**Comment by russjones** [comment]
@vitorenesduarte This is another issue, before we start hacking on this, let's scope out the work to see how much work it will be before we schedule it.

**Comment by vitorenesduarte** [comment]
_(from @fspmarshall via DM)_

### Background

Teleport used to treat every individual service running on the instance as if it were a separate teleport process. Each one got its own client to the auth server with its own certificate specific to that service (system role).  We're trying to move away from that model for various reasons.

As part of that work, a new `Instance` system role was added a while back, which is implicitly held by any teleport instance that has at least one service running on it (`Auth`, `Node`, etc). Previously, each service had a separate certificate specifically for its system role. The new `Instance` system role's certificate encodes a list of all the system roles that the instance is allowed to hold. The idea being that instead of having, for example, one client with a `Node` cert, and one client with a `Database` cert, you could have a single `Instance` client whose cert listed all the system roles.

We added the new `Instance` client ([here](https://github.com/gravitational/teleport/blob/8fe2a74d0263173a5c2039d43be1a342573743f6/lib/service/service.go#L326-L328)), but we couldn't just immediately start using it in place of per-service clients. First, because we didn't want the new feature to brick existing services if it had bugs, but also because the process of getting the `Instance` cert on existing instances was kind of complex. If the instance still has a valid join token, then you can just ask the auth server for an `Instance` cert, and it will create one with all of the roles granted by the token. The problem is that many existing instances have had their tokens expire a long time ago. In that case, we needed the instance to go through a process of "proving" which system roles it actually held, using the old per-service clients. This process ([here](https://github.com/gravitational/teleport/blob/8fe2a74d0263173a5c2039d43be1a342573743f6/lib/service/connect.go#L144-L169)) was multi-step, and potentially slow if lots of instances were trying to do it at the same time (lots of backend load), so we made the decision to have the instance clients be created lazily (i.e. let the existing services start with their normal per-service client).

Due to the laziness of the instance client, it is currently only used for new features (the inventory control stream, and the v2 heartbeats, which only the `Node` role uses at the moment). That way the services can start up without waiting for the `Instance` cert to be granted.

`Instance` certs were added in `v10`, so clients on `v11` onward should have instance certs immediately (either because they acquired it via assertions when they were on `v10`, or because they are new instances and can get it via their join token). This means we should be able to remove the assertion logic, and it should be safe to force services to wait for the instance client to become healthy before starting up. Neither of these things have actually been done yet. No one has had any reason to do it since only new features rely on the `Instance` client.

### Current problem: extra connection

The current issue is that there are two connections to auth. This is because we're still using the old per-service clients for most stuff, and the `Instance` client only does some basic heartbeating of instance state.

The solution is to deprecate the old per-service clients, and use the `Instance` client in their stead (e.g. [here](https://github.com/gravitational/teleport/blob/8fe2a74d0263173a5c2039d43be1a342573743f6/lib/service/service.go#L2143-L2157) instead of waiting for an SSH/Node client and then using that for all of the ssh service setup, we'd just wait for the instance client to become available and use that).

The expectation is that this will "just work" (i.e. no changes will need to be made to the auth server), since the auth servers RBAC logic was updated at the time we added the `Instance` cert, so that, for example, an `Instance` cert that lists roles `Node` and `Database` has the capabilities of both a `Node` cert and a `Database` cert. In practice, it wouldn't be surprising if we find edge-cases that need fixing when we actually migrate the clients.

Note that we still need the per-service certs for stuff other than auth clients (e.g. we still create a separate proxy reversetunnel for each service), we just want to get rid of the per-service auth clients.

Also, while we want to migrate all of them, just migrating the SSH (`Node`) client is probably all we actually need to do for now. The vast majority of all teleport instances are just nodes (clusters can have tens of thousands of nodes, compared to typically less than 100 for most other services).

No deep dive investigating the best way to change the existing `lib/service` logic to use the instance client everywhere has been done. We probably want to continue having separate asynchronous "Identity events" for each service, and simply change it so that instead of having each one contain a different client, they all contain a reference to the instance client.

**Comment by zmb3** [comment]
One place that could benefit from this new instance client is the upload completer. Each teleport process runs 1 uploader completer service, it doesn't matter how many other Teleport services are running in the process. 

In this case, we don't really care which client we use, we just need any old client, so we pick the first one we can find:  

https://github.com/gravitational/teleport/blob/0f21a8b8e376018fa0d9b96892671e98a661693f/lib/service/service.go#L2463-L2471

Once we can be sure there will always be an instance client, we should use that one instead.  It would allow us to get rid of this complicated test that ensures that whichever client we pick has the correct roles: 

https://github.com/gravitational/teleport/blob/0f21a8b8e376018fa0d9b96892671e98a661693f/lib/auth/auth_with_roles_test.go#L3579-L3583
